README.md
<p align="center">
  <img src="https://static.hoa-project.net/Image/Hoa.svg" alt="Hoa" width="250px" />
</p>

---

<p align="center">
  <a href="https://travis-ci.org/hoaproject/console"><img src="https://img.shields.io/travis/hoaproject/console/master.svg" alt="Build status" /></a>
  <a href="https://coveralls.io/github/hoaproject/console?branch=master"><img src="https://img.shields.io/coveralls/hoaproject/console/master.svg" alt="Code coverage" /></a>
  <a href="https://packagist.org/packages/hoa/console"><img src="https://img.shields.io/packagist/dt/hoa/console.svg" alt="Packagist" /></a>
  <a href="https://hoa-project.net/LICENSE"><img src="https://img.shields.io/packagist/l/hoa/console.svg" alt="License" /></a>
</p>
<p align="center">
  Hoa is a <strong>modular</strong>, <strong>extensible</strong> and
  <strong>structured</strong> set of PHP libraries.<br />
  Moreover, Hoa aims at being a bridge between industrial and research worlds.
</p>

# Hoa\Console

[![Help on IRC](https://img.shields.io/badge/help-%23hoaproject-ff0066.svg)](https://webchat.freenode.net/?channels=#hoaproject)
[![Help on Gitter](https://img.shields.io/badge/help-gitter-ff0066.svg)](https://gitter.im/hoaproject/central)
[![Documentation](https://img.shields.io/badge/documentation-hack_book-ff0066.svg)](https://central.hoa-project.net/Documentation/Library/Console)
[![Board](https://img.shields.io/badge/organisation-board-ff0066.svg)](https://waffle.io/hoaproject/console)

This library allows to interact easily with a terminal: getoption, cursor,
window, processus, readline etc.

[Learn more](https://central.hoa-project.net/Documentation/Library/Console).

## Installation

With [Composer](https://getcomposer.org/), to include this library into
your dependencies, you need to
require [`hoa/console`](https://packagist.org/packages/hoa/console):

```sh
$ composer require hoa/console '~3.0'
```

For more installation procedures, please read [the Source
page](https://hoa-project.net/Source.html).

## Testing

Before running the test suites, the development dependencies must be installed:

```sh
$ composer install
```

Then, to run all the test suites:

```sh
$ vendor/bin/hoa test:run
```

For more information, please read the [contributor
guide](https://hoa-project.net/Literature/Contributor/Guide.html).

## Quick usage

We propose a quick overview of some features: cursor, window, readline,
processus and finally getoption.

### Cursor

The `Hoa\Console\Cursor` class allows to manipulate the cursor. Here is a list
of some operations:

  * `move`,
  * `moveTo`,
  * `save`,
  * `restore`,
  * `clear`,
  * `hide`,
  * `show`,
  * `getPosition`,
  * `colorize`,
  * etc.

The API is very straightforward. For example, we can use `l`, `left` or `←` to
move the cursor on the left column. Thus we move the cursor to the left 3 times
and then to the top 2 times:

```php
Hoa\Console\Cursor::move('left left left up up');
```

… or with Unicode symbols:

```php
Hoa\Console\Cursor::move('← ← ← ↑ ↑');
```

This method moves the cursor relatively from its current position, but we are
able to move the cursor to absolute coordinates:

```php
Hoa\Console\Cursor::moveTo(13, 42);
```

We are also able to save the current cursor position, to move, clear etc., and
then to restore the saved position:

```php
Hoa\Console\Cursor::save();     // save
Hoa\Console\Cursor::move('↓');  // move below
Hoa\Console\Cursor::clear('↔'); // clear the line
echo 'Something below…';        // write something
Hoa\Console\Cursor::restore();  // restore
```

Another example with colors:

```php
Hoa\Console\Cursor::colorize(
    'underlined foreground(yellow) background(#932e2e)'
);
```

Please, read the API documentation for more informations.

### Mouse

The `Hoa\Console\Mouse` class allows to listen the mouse actions and provides
the following listeners: `mouseup`, `mousedown`, `wheelup` and `wheeldown`.
Example:

```php
$mouse = Hoa\Console\Mouse::getInstance();
$mouse->on('mousedown', function ($bucket) {
    print_r($bucket->getData());
});

$mouse::track();
```

And then, when we left-click, we will see:

```
Array
(
    [x] => 69
    [y] => 30
    [button] => left
    [shift] =>
    [meta] =>
    [ctrl] =>
)
```

When we left-click while hiting the shift key, we will see:

```
Array
(
    [x] => 71
    [y] => 32
    [button] => left
    [shift] => 1
    [meta] =>
    [ctrl] =>
)
```

This is an experimental API.

### Window

The `Hoa\Console\Window` class allows to manipulate the window. Here is a list
of some operations:

  * `setSize`,
  * `getSize`,
  * `moveTo`,
  * `getPosition`,
  * `scroll`,
  * `minimize`,
  * `restore`,
  * `raise`,
  * `setTitle`,
  * `getTitle`,
  * `copy`,
  * etc.

Furthermore, we have the `hoa://Event/Console/Window:resize` event channel to
listen when the window has been resized.

For example, we resize the window to 40 lines and 80 columns, and then we move
the window to 400px horizontally and 100px vertically:

```php
Hoa\Console\Window::setSize(40, 80);
Hoa\Console\Window::moveTo(400, 100);
```

If we do not like our user, we are able to minimize its window:

```php
Hoa\Console\Window::minimize();
sleep(2);
Hoa\Console\Window::restore();
```

We are also able to set or get the title of the window:

```php
Hoa\Console\Window::setTitle('My awesome application');
```

Finally, if we have a complex application layout, we can repaint it when the
window is resized by listening the `hoa://Event/Console/Window:resize` event
channel:

```php
Hoa\Event\Event::getEvent('hoa://Event/Console/Window:resize')
    ->attach(function (Hoa\Event\Bucket $bucket) {
        $data = $bucket->getData();
        $size = $data['size'];

        echo
            'New dimensions: ', $size['x'], ' lines x ',
            $size['y'], ' columns.', "\n";
    });
```

Please, read the API documentation for more informations

### Readline

The `Hoa\Console\Readline\Readline` class provides an advanced readline which
allows the following operations:

  * edition,
  * history,
  * autocompletion.

It supports UTF-8. It is based on bindings, and here are some:

  * `arrow up` and `arrow down`: move in the history,
  * `arrow left` and `arrow right`: move the cursor left and right,
  * `Ctrl-A`: move to the beginning of the line,
  * `Ctrl-E`: move to the end of the line,
  * `Ctrl-B`: move backward of one word,
  * `Ctrl-F`: move forward of one word,
  * `Ctrl-W`: delete first backard word,
  * `Backspace`: delete first backward character,
  * `Enter`: submit the line,
  * `Tab`: autocomplete.

Thus, to read one line:

```php
$readline = new Hoa\Console\Readline\Readline();
$line     = $readline->readLine('> '); // “> ” is the prefix of the line.
```

The `Hoa\Console\Readline\Password` allows the same operations but without
printing on STDOUT.

```php
$password = new Hoa\Console\Readline\Password();
$line     = $password->readLine('password: ');
```

We are able to add a mapping with the help of the
`Hoa\Console\Readline\Readline::addMapping` method. We use `\e[…` for `\033[`,
`\C-…` for `Ctrl-…` and a character for the rest. We can associate a character
or a callable:

```php
$readline->addMapping('a', 'z'); // crazy, we replace “a” by “z”.
$readline->addMapping('\C-R', function ($readline) {
    // do something when pressing Ctrl-R.
});
```

We are also able to manipulate the history, thanks to the `addHistory`,
`clearHistory`, `getHistory`, `previousHistory` and `nextHistory` methods on the
`Hoa\Console\Readline\Readline` class.

Finally, we have autocompleters that are enabled on `Tab`. If one solution is
proposed, it will be inserted directly. If many solutions are proposed, we are
able to navigate in a menu to select the solution (with the help of keyboard
arrows, Enter, Esc etc.). Also, we are able to combine autocompleters. The
following example combine the `Word` and `Path` autocompleters:

```php
$functions = get_defined_functions();
$readline->setAutocompleter(
    new Hoa\Console\Readline\Autocompleter\Aggregate([
        new Hoa\Console\Readline\Autocompleter\Path(),
        new Hoa\Console\Readline\Autocompleter\Word($functions['internal'])
    ])
);
```

Here is an example of the result:

![Autocompleters in action](http://central.hoa-project.net/Resource/Library/Console/Documentation/Image/Readline_autocompleters.gif?format=raw)

On Windows, a readline is equivalent to a simple `fgets(STDIN)`.

### Processus

The `Hoa\Console\Processus` class allows to manipulate processus as a stream
which implements `Hoa\Stream\IStream\In`, `Hoa\Stream\IStream\Out` and
`Hoa\Stream\IStream\Pathable` interfaces (please, see the
[`Hoa\Stream` library](http://central.hoa-project.net/Resource/Library/Stream)).

Basically, we can read STDOUT like this:

```php
$processus = new Hoa\Console\Processus('ls');
$processus->open();
echo $processus->readAll();
```

And we can write on STDIN like this:

```php
$processus->writeAll('foobar');
```

etc. This is very classical.

`Hoa\Console\Processus` also proposes many events: `start`, `stop`, `input`,
`output` and `timeout`. Thus:

```php
$processus = new Hoa\Console\Processus('ls');
$processus->on('output', function (Hoa\Event\Bucket $bucket) {
    $data = $bucket->getData();
    echo '> ', $data['line'], "\n";
});
$processus->run();
```

We are also able to read and write on more pipes than 0 (STDOUT), 1 (STDIN) and
2 (STDERR). In the same way, we can set the current working directory of the
processus and its environment.

We can quickly execute a processus without using a stream with the help of the

CHANGELOG.md
# 3.17.01.11

  * Quality: Happy new year! (Alexis von Glasow, 2017-01-10T12:48:50+01:00)

# 3.16.11.08

  * Documentation: Add the “Learn more” link. (Ivan Enderlin, 2016-10-14T23:40:05+02:00)
  * Documentation: New `README.md` file. (Ivan Enderlin, 2016-10-14T23:29:22+02:00)
  * Documentation: Fix `docs` and `source` links. (Ivan Enderlin, 2016-10-05T20:28:46+02:00)
  * Documentation: Update `support` properties. (Ivan Enderlin, 2016-10-05T20:12:17+02:00)
  * Documentation: Use TLS on `central.hoa`. (Ivan Enderlin, 2016-09-09T14:59:03+02:00)

# 3.16.09.06

  * Documentation: Fix API documentation. (Alexis von Glasow, 2016-07-11T00:01:57+02:00)
  * Quality: Fix example CS in the `README.md`. (Ivan Enderlin, 2016-07-01T16:37:50+02:00)
  * Autocompleter: Force to work on a sub-line. (Ivan Enderlin, 2016-05-22T15:46:58+02:00)
  * Quality: Fix methods ordering. (Ivan Enderlin, 2016-02-24T07:28:41+01:00)
  * Implement `Hoa\Console\Output::getStream` which is now required by `Hoa\Stream\IStream\Out`. (Metalaka, 2015-11-01T20:15:43+01:00)
  * Fix phpDoc. (Metalaka, 2015-11-01T20:13:12+01:00)

# 3.16.01.14

  * Composer: New stable libraries. (Ivan Enderlin, 2016-01-14T21:45:35+01:00)

# 3.16.01.11

  * Quality: Drop PHP5.4. (Ivan Enderlin, 2016-01-11T09:15:26+01:00)
  * Quality: Run devtools:cs. (Ivan Enderlin, 2016-01-09T08:59:18+01:00)
  * Core: Remove `Hoa\Core`. (Ivan Enderlin, 2016-01-09T08:08:44+01:00)
  * Consistency: Update `registerShutdownFunction`. (Ivan Enderlin, 2015-12-09T06:58:00+01:00)
  * Consistency: Remove `from` calls. (Ivan Enderlin, 2015-12-09T06:35:35+01:00)
  * Consistency: Update `registerShutdownFunction`. (Ivan Enderlin, 2015-12-08T23:41:11+01:00)
  * Event: Remove `event` calls. (Ivan Enderlin, 2015-12-08T22:48:06+01:00)
  * Consistency: Use `Hoa\Consistency`. (Ivan Enderlin, 2015-12-08T11:01:36+01:00)
  * Event: Use `Hoa\Event`. (Ivan Enderlin, 2015-11-23T21:48:33+01:00)
  * Exception: Use `Hoa\Exception`. (Ivan Enderlin, 2015-11-20T07:17:38+01:00)
  * Test: Add specificity for Windows. (Ivan Enderlin, 2015-11-10T13:29:22+01:00)
  * Test: Write test suite for `…completer\Aggregate`. (Ivan Enderlin, 2015-11-10T08:45:10+01:00)
  * Test: Write test suite for `…e\Readline\Password`. (Ivan Enderlin, 2015-10-29T23:00:01+01:00)
  * Test: Use `beforeTestMethod` instead of `setUp`. (Ivan Enderlin, 2015-10-29T13:41:12+01:00)
  * Terminfo: Add the `xterm-256color` database. (Ivan Enderlin, 2015-10-29T13:40:30+01:00)
  * Test: Write test suite for `…Autocompleter\Path`. (Ivan Enderlin, 2015-10-29T08:46:04+01:00)
  * Test: Write test suite for `…Autocompleter\Word`. (Ivan Enderlin, 2015-10-29T08:10:47+01:00)
  * Readline: Use `Console::getInput`. (Ivan Enderlin, 2015-10-29T07:38:13+01:00)
  * Test: Write test suite for `Hoa\Console\GetOption`. (Ivan Enderlin, 2015-10-29T07:37:57+01:00)
  * GetOption: Reset the `$optionValue` all the time. (Ivan Enderlin, 2015-10-29T07:37:37+01:00)
  * CS: Clean namespaces and fix some styles. (Ivan Enderlin, 2015-10-28T07:56:40+01:00)
  * Test: Write test suite for `Hoa\Console\Mouse`. (Ivan Enderlin, 2015-10-28T07:54:38+01:00)
  * Mouse: Untrack when tracking fails. (Ivan Enderlin, 2015-10-28T07:52:48+01:00)
  * Mouse: New constants representing pointer codes. (Ivan Enderlin, 2015-10-28T07:50:50+01:00)
  * Test: Write test suite for `Hoa\Console\Window`. (Ivan Enderlin, 2015-10-27T10:12:44+01:00)
  * Window: The constructor must be private. (Ivan Enderlin, 2015-10-27T09:32:54+01:00)
  * Test: Write test suite for `Hoa\Console\Cursor`. (Ivan Enderlin, 2015-10-26T13:37:53+01:00)
  * Test: Write test suite for `Hoa\Console\Tput`. (Ivan Enderlin, 2015-10-26T13:37:33+01:00)
  * Test: Write test suite for `Hoa\Console\Parser`. (Ivan Enderlin, 2015-10-26T13:37:13+01:00)
  * Test: Write test suite for `Hoa\Console`. (Ivan Enderlin, 2015-10-26T13:36:42+01:00)
  * Console: Add the `setTput` static method. (Ivan Enderlin, 2015-10-26T13:34:37+01:00)
  * Console: Ensure `STDIN` is defined before using it. (Ivan Enderlin, 2015-10-26T13:34:02+01:00)
  * Tput: If no terminfo found, fallback to `xterm`. (Ivan Enderlin, 2015-10-28T14:07:18+01:00)
  * Tput: Check that `TERM` is set **and** not empty. (Ivan Enderlin, 2015-10-28T14:06:36+01:00)
  * Update API documentation. (Ivan Enderlin, 2015-10-26T08:39:15+01:00)
  * Console: Solve `stty -f` vs. `stty -F` issue. (Ivan Enderlin, 2015-10-22T09:29:42+02:00)
  * Input: Add the `getStream` method. (Ivan Enderlin, 2015-10-22T08:34:02+02:00)
  * Test the `Input` class. (Ivan Enderlin, 2015-10-21T17:04:52+02:00)
  * Create the `Input` interface. (Ivan Enderlin, 2015-09-24T07:44:50+02:00)
  * `stty` uses `/dev/tty` instead of the std. input. (Ivan Enderlin, 2015-09-24T07:16:48+02:00)

Cursor.php
<?php

/**
 * Hoa
 *
 *
 * @license
 *
 * New BSD License
 *
 * Copyright © 2007-2016, Hoa community. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the Hoa nor the names of its contributors may be
 *       used to endorse or promote products derived from this software without
 *       specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

namespace Hoa\Console;

/**
 * Class \Hoa\Console\Cursor.
 *
 * Allow to manipulate the cursor.
 *
 * @copyright  Copyright © 2007-2016 Hoa community
 * @license    New BSD License
 */
class Cursor
{
    /**
     * Move the cursor.
     * Steps can be:
     *     • u, up,    ↑ : move to the previous line;
     *     • U, UP       : move to the first line;
     *     • r, right, → : move to the next column;
     *     • R, RIGHT    : move to the last column;
     *     • d, down,  ↓ : move to the next line;
     *     • D, DOWN     : move to the last line;
     *     • l, left,  ← : move to the previous column;
     *     • L, LEFT     : move to the first column.
     * Steps can be concatened by a single space if $repeat is equal to 1.
     *
     * @param   string  $steps     Steps.
     * @param   int     $repeat    How many times do we move?
     * @return  void
     */
    public static function move($steps, $repeat = 1)
    {
        if (1 > $repeat) {
            return;
        } elseif (1 === $repeat) {
            $handle = explode(' ', $steps);
        } else {
            $handle = explode(' ', $steps, 1);
        }

        $tput   = Console::getTput();
        $output = Console::getOutput();

        foreach ($handle as $step) {
            switch ($step) {
                case 'u':
                case 'up':
                case '↑':
                    $output->writeAll(
                        str_replace(
                            '%p1%d',
                            $repeat,
                            $tput->get('parm_up_cursor')
                        )
                    );

                    break;

                case 'U':
                case 'UP':
                    static::moveTo(null, 1);

                    break;

                case 'r':
                case 'right':
                case '→':
                    $output->writeAll(
                        str_replace(
                            '%p1%d',
                            $repeat,
                            $tput->get('parm_right_cursor')
                        )
                    );

                    break;

                case 'R':
                case 'RIGHT':
                    static::moveTo(9999);

                    break;

                case 'd':
                case 'down':
                case '↓':
                    $output->writeAll(
                        str_replace(
                            '%p1%d',
                            $repeat,
                            $tput->get('parm_down_cursor')
                        )
                    );

                    break;

                case 'D':
                case 'DOWN':
                    static::moveTo(null, 9999);

                    break;

                case 'l':
                case 'left':
                case '←':
                    $output->writeAll(
                        str_replace(
                            '%p1%d',
                            $repeat,
                            $tput->get('parm_left_cursor')
                        )
                    );

                    break;

                case 'L':
                case 'LEFT':
                    static::moveTo(1);

                    break;
            }
        }

        return;
    }

    /**
     * Move to the line X and the column Y.
     * If null, use the current coordinate.
     *
     * @param   int  $x    X coordinate.
     * @param   int  $y    Y coordinate.
     * @return  void
     */
    public static function moveTo($x = null, $y = null)
    {
        if (null === $x || null === $y) {
            $position = static::getPosition();

            if (null === $x) {
                $x = $position['x'];
            }

            if (null === $y) {
                $y = $position['y'];
            }
        }

        Console::getOutput()->writeAll(
            str_replace(
                ['%i%p1%d', '%p2%d'],
                [$y, $x],
                Console::getTput()->get('cursor_address')
            )
        );

        return;
    }

    /**
     * Get current position (x and y) of the cursor.
     *
     * @return  array
     */
    public static function getPosition()
    {
        $tput  = Console::getTput();
        $user7 = $tput->get('user7');

        if (null === $user7) {
            return [
                'x' => 0,
                'y' => 0
            ];
        }

        Console::getOutput()->writeAll($user7);

        $input = Console::getInput();

        // Read $tput->get('user6').
        $input->read(2); // skip \033 and [.

        $x      = null;
        $y      = null;
        $handle = &$y;

        do {
            $char = $input->readCharacter();

            switch ($char) {
                case ';':
                    $handle = &$x;

                    break;

                case 'R':
                    break 2;

                default:
                    $handle .= $char;
            }
        } while (true);

        return [
            'x' => (int) $x,
            'y' => (int) $y
        ];
    }

    /**
     * Save current position.
     *
     * @return  void
     */
    public static function save()
    {
        Console::getOutput()->writeAll(
            Console::getTput()->get('save_cursor')
        );

        return;
    }

    /**
     * Restore cursor to the last saved position.
     *
     * @return  void
     */
    public static function restore()
    {
        Console::getOutput()->writeAll(
            Console::getTput()->get('restore_cursor')
        );

        return;
    }

    /**
     * Clear the screen.
     * Part can be:
     *     • a, all,   ↕ : clear entire screen and static::move(1, 1);
     *     • u, up,    ↑ : clear from cursor to beginning of the screen;
     *     • r, right, → : clear from cursor to the end of the line;
     *     • d, down,  ↓ : clear from cursor to end of the screen;
     *     • l, left,  ← : clear from cursor to beginning of the screen;
     *     •    line,  ↔ : clear all the line and static::move(1).
     * Parts can be concatenated by a single space.
     *
     * @param   string  $parts    Parts to clean.
     * @return  void
     */
    public static function clear($parts = 'all')
    {
        $tput   = Console::getTput();
        $output = Console::getOutput();

        foreach (explode(' ', $parts) as $part) {
            switch ($part) {
                case 'a':
                case 'all':
                case '↕':
                    $output->writeAll($tput->get('clear_screen'));
                    static::moveTo(1, 1);

                    break;

                case 'u':
                case 'up':
                case '↑':
                    $output->writeAll("\033[1J");

                    break;

                case 'r':
                case 'right':
                case '→':
                    $output->writeAll($tput->get('clr_eol'));

                    break;

                case 'd':
                case 'down':
                case '↓':
                    $output->writeAll($tput->get('clr_eos'));

                    break;

                case 'l':
                case 'left':
                case '←':
                    $output->writeAll($tput->get('clr_bol'));

                    break;

                case 'line':
                case '↔':
                    $output->writeAll("\r" . $tput->get('clr_eol'));

                    break;
            }
        }

        return;
    }

    /**
     * Hide the cursor.
     *
     * @return  void
     */
    public static function hide()
    {
        Console::getOutput()->writeAll(
            Console::getTput()->get('cursor_invisible')
        );

        return;
    }

    /**
     * Show the cursor.
     *
     * @return  void
     */
    public static function show()
    {
        Console::getOutput()->writeAll(
            Console::getTput()->get('cursor_visible')
        );

        return;
    }

    /**
     * Colorize cursor.
     * Attributes can be:
     *     •  n,         normal           : normal;
     *     •  b,         bold             : bold;
     *     •  u,         underlined       : underlined;
     *     •  bl,        blink            : blink;
     *     •  i,         inverse          : inverse;
     *     • !b,        !bold             : normal weight;
     *     • !u,        !underlined       : not underlined;
     *     • !bl,       !blink            : steady;
     *     • !i,        !inverse          : positive;
     *     • fg(color), foreground(color) : set foreground to “color”;
     *     • bg(color), background(color) : set background to “color”.
     * “color” can be:
     *     • default;
     *     • black;
     *     • red;
     *     • green;
     *     • yellow;
     *     • blue;
     *     • magenta;
     *     • cyan;
     *     • white;
     *     • 0-256 (classic palette);
     *     • #hexa.
     * Attributes can be concatenated by a single space.
     *
     * @param   string  $attributes    Attributes.
     * @return  void
     */
    public static function colorize($attributes)
    {
        static $_rgbTo256 = null;

        if (null === $_rgbTo256) {
            $_rgbTo256 = [
                '000000', '800000', '008000', '808000', '000080', '800080',
                '008080', 'c0c0c0', '808080', 'ff0000', '00ff00', 'ffff00',
                '0000ff', 'ff00ff', '00ffff', 'ffffff', '000000', '00005f',
                '000087', '0000af', '0000d7', '0000ff', '005f00', '005f5f',
                '005f87', '005faf', '005fd7', '005fff', '008700', '00875f',
                '008787', '0087af', '0087d7', '0087ff', '00af00', '00af5f',
                '00af87', '00afaf', '00afd7', '00afff', '00d700', '00d75f',
                '00d787', '00d7af', '00d7d7', '00d7ff', '00ff00', '00ff5f',
                '00ff87', '00ffaf', '00ffd7', '00ffff', '5f0000', '5f005f',
                '5f0087', '5f00af', '5f00d7', '5f00ff', '5f5f00', '5f5f5f',
                '5f5f87', '5f5faf', '5f5fd7', '5f5fff', '5f8700', '5f875f',
                '5f8787', '5f87af', '5f87d7', '5f87ff', '5faf00', '5faf5f',
                '5faf87', '5fafaf', '5fafd7', '5fafff', '5fd700', '5fd75f',
                '5fd787', '5fd7af', '5fd7d7', '5fd7ff', '5fff00', '5fff5f',
                '5fff87', '5fffaf', '5fffd7', '5fffff', '870000', '87005f',
                '870087', '8700af', '8700d7', '8700ff', '875f00', '875f5f',
                '875f87', '875faf', '875fd7', '875fff', '878700', '87875f',
                '878787', '8787af', '8787d7', '8787ff', '87af00', '87af5f',
                '87af87', '87afaf', '87afd7', '87afff', '87d700', '87d75f',
                '87d787', '87d7af', '87d7d7', '87d7ff', '87ff00', '87ff5f',
                '87ff87', '87ffaf', '87ffd7', '87ffff', 'af0000', 'af005f',
                'af0087', 'af00af', 'af00d7', 'af00ff', 'af5f00', 'af5f5f',
                'af5f87', 'af5faf', 'af5fd7', 'af5fff', 'af8700', 'af875f',
                'af8787', 'af87af', 'af87d7', 'af87ff', 'afaf00', 'afaf5f',
                'afaf87', 'afafaf', 'afafd7', 'afafff', 'afd700', 'afd75f',
                'afd787', 'afd7af', 'afd7d7', 'afd7ff', 'afff00', 'afff5f',
                'afff87', 'afffaf', 'afffd7', 'afffff', 'd70000', 'd7005f',
                'd70087', 'd700af', 'd700d7', 'd700ff', 'd75f00', 'd75f5f',
                'd75f87', 'd75faf', 'd75fd7', 'd75fff', 'd78700', 'd7875f',
                'd78787', 'd787af', 'd787d7', 'd787ff', 'd7af00', 'd7af5f',
                'd7af87', 'd7afaf', 'd7afd7', 'd7afff', 'd7d700', 'd7d75f',
                'd7d787', 'd7d7af', 'd7d7d7', 'd7d7ff', 'd7ff00', 'd7ff5f',
                'd7ff87', 'd7ffaf', 'd7ffd7', 'd7ffff', 'ff0000', 'ff005f',
                'ff0087', 'ff00af', 'ff00d7', 'ff00ff', 'ff5f00', 'ff5f5f',
                'ff5f87', 'ff5faf', 'ff5fd7', 'ff5fff', 'ff8700', 'ff875f',
                'ff8787', 'ff87af', 'ff87d7', 'ff87ff', 'ffaf00', 'ffaf5f',
                'ffaf87', 'ffafaf', 'ffafd7', 'ffafff', 'ffd700', 'ffd75f',
                'ffd787', 'ffd7af', 'ffd7d7', 'ffd7ff', 'ffff00', 'ffff5f',
                'ffff87', 'ffffaf', 'ffffd7', 'ffffff', '080808', '121212',
                '1c1c1c', '262626', '303030', '3a3a3a', '444444', '4e4e4e',
                '585858', '606060', '666666', '767676', '808080', '8a8a8a',
                '949494', '9e9e9e', 'a8a8a8', 'b2b2b2', 'bcbcbc', 'c6c6c6',
                'd0d0d0', 'dadada', 'e4e4e4', 'eeeeee'
            ];
        }

        $tput = Console::getTput();

        if (1 >= $tput->count('max_colors')) {
            return;
        }

        $handle = [];

        foreach (explode(' ', $attributes) as $attribute) {
            switch ($attribute) {
                case 'n':
                case 'normal':
                    $handle[] = 0;

                    break;

                case 'b':
                case 'bold':
                    $handle[] = 1;

                    break;

                case 'u':
                case 'underlined':
                    $handle[] = 4;

                    break;

                case 'bl':
                case 'blink':
                    $handle[] = 5;

                    break;

                case 'i':
                case 'inverse':
                    $handle[] = 7;

                    break;

                case '!b':
                case '!bold':
                    $handle[] = 22;

                    break;

                case '!u':
                case '!underlined':
                    $handle[] = 24;

                    break;

                case '!bl':
                case '!blink':
                    $handle[] = 25;

                    break;

                case '!i':
                case '!inverse':
                    $handle[] = 27;

                    break;

                default:
                    if (0 === preg_match('#^([^\(]+)\(([^\)]+)\)$#', $attribute, $m)) {
                        break;
                    }

                    $shift = 0;

                    switch ($m[1]) {
                        case 'fg':
                        case 'foreground':
                            $shift = 0;

                            break;

                        case 'bg':
                        case 'background':
                            $shift = 10;

                            break;

                        default:
                            break 2;
                    }

                    $_handle  = 0;
                    $_keyword = true;

                    switch ($m[2]) {
                        case 'black':
                            $_handle = 30;

                            break;

                        case 'red':
                            $_handle = 31;

                            break;

                        case 'green':
                            $_handle = 32;

                            break;

                        case 'yellow':
                            $_handle = 33;

                            break;

                        case 'blue':
                            $_handle = 34;

                            break;

                        case 'magenta':
                            $_handle = 35;

                            break;

                        case 'cyan':
                            $_handle = 36;

                            break;

                        case 'white':
                            $_handle = 37;

                            break;

                        case 'default':
                            $_handle = 39;

                            break;

                        default:
                            $_keyword = false;

                            if (256 <=  $tput->count('max_colors') &&
                                '#' === $m[2][0]) {
                                $rgb      = hexdec(substr($m[2], 1));
                                $r        = ($rgb >> 16) & 255;
                                $g        = ($rgb >>  8) & 255;
                                $b        =  $rgb        & 255;
                                $distance = null;

                                foreach ($_rgbTo256 as $i => $_rgb) {
                                    $_rgb = hexdec($_rgb);
                                    $_r   = ($_rgb >> 16) & 255;
                                    $_g   = ($_rgb >>  8) & 255;
                                    $_b   =  $_rgb        & 255;

                                    $d = sqrt(
                                        pow($_r - $r, 2)
                                      + pow($_g - $g, 2)
                                      + pow($_b - $b, 2)
                                    );

                                    if (null === $distance ||
                                        $d   <=  $distance) {
                                        $distance = $d;
                                        $_handle  = $i;
                                    }
                                }
                            } else {
                                $_handle = intval($m[2]);
                            }
                    }

                    if (true === $_keyword) {
                        $handle[] = $_handle + $shift;
                    } else {
                        $handle[] = (38 + $shift) . ';5;' . $_handle;
                    }
            }
        }

        Console::getOutput()->writeAll("\033[" . implode(';', $handle) . "m");

        return;
    }

    /**
     * Change color number to a specific RGB color.
     *
     * @param   int  $fromCode    Color number.
     * @param   int  $toColor     RGB color.
     * @return  void
     */
    public static function changeColor($fromCode, $toColor)
    {
        $tput = Console::getTput();

        if (true !== $tput->has('can_change')) {
            return;
        }

        $r = ($toColor >> 16) & 255;
        $g = ($toColor >>  8) & 255;
        $b =  $toColor        & 255;

        Console::getOutput()->writeAll(
            str_replace(
                [
                    '%p1%d',
                    'rgb:',
                    '%p2%{255}%*%{1000}%/%2.2X/',
                    '%p3%{255}%*%{1000}%/%2.2X/',
                    '%p4%{255}%*%{1000}%/%2.2X'
                ],
                [
                    $fromCode,
                    '',
                    sprintf('%02x', $r),
                    sprintf('%02x', $g),
                    sprintf('%02x', $b)
                ],
                $tput->get('initialize_color')
            )
        );

        return;
    }

    /**
     * Set cursor style.
     * Style can be:
     *     • b, block,     ▋: block;
     *     • u, underline, _: underline;
     *     • v, vertical,  |: vertical.
     *
     * @param   int   $style    Style.
     * @param   bool  $blink    Whether the cursor is blink or steady.
     * @return  void
     */
    public static function setStyle($style, $blink = true)
    {
        if (OS_WIN) {
            return;
        }

        switch ($style) {
            case 'b':
            case 'block':
            case '▋':
                $_style = 1;

                break;

            case 'u':
            case 'underline':
            case '_':
                $_style = 2;

                break;

            case 'v':
            case 'vertical':
            case '|':
                $_style = 5;

                break;
        }

        if (false === $blink) {
            ++$_style;
        }

        // Not sure what tput entry we can use here…
        Console::getOutput()->writeAll("\033[" . $_style . " q");

        return;
    }

    /**
     * Make a stupid “bip”.
     *
     * @return  void
     */
    public static function bip()
    {
        Console::getOutput()->writeAll(
            Console::getTput()->get('bell')
        );

        return;
    }
}

/**
 * Advanced interaction.
 */
Console::advancedInteraction();
